vyper/ast/folding.py

```diff
@@ -198,18 +208,27 @@ def replace_user_defined_constants(vyper_module: vy_ast.Module) -> int:
 # TODO constant folding on log events
 
 
-def _replace(old_node, new_node, type_=None):
+def _replace(old_node, new_node, var_info=None):
     if isinstance(new_node, vy_ast.Constant):
         new_node = new_node.from_node(old_node, value=new_node.value)
-        if type_:
-            new_node._metadata["type"] = type_
+        if var_info is not None:
+            new_node._metadata["varinfo"] = var_info
         return new_node
     elif isinstance(new_node, vy_ast.List):
-        base_type = type_.value_type if type_ else None
-        list_values = [_replace(old_node, i, type_=base_type) for i in new_node.elements]
+        base_type_varinfo = None
+        if var_info is not None:
+            base_type_varinfo = VarInfo(
+                typ=var_info.typ.value_type,
+                location=DataLocation.CODE,
+                is_constant=var_info.is_constant,
+                is_public=var_info.is_public,
+                is_immutable=var_info.is_immutable,
+                decl_node=var_info.decl_node,
+            )
```

Comment:
fyi - when i kind of complicated code inlined like this, i'd prefer to see a helper function to do this which is closer to the definition of `VarInfo`, it's too easy to update the fields of `VarInfo` and then forget to update all uses of the constructor. in particular, this looks like a good use case for `VarInfo.copy_with_type(type_)`. but as i mentioned in https://github.com/vyperlang/vyper/pull/3189#pullrequestreview-1427258268, we probably want to use ExprInfo instead anyway.